styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -366,41 +366,61 @@ private synchronized void tryPollPods() {
 
     for (Pod pod : list.getItems()) {
       logEvent(Watcher.Action.MODIFIED, pod, resourceVersion, true);
-      inspectPod(Watcher.Action.MODIFIED, pod);
+      final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
+      if (!workflowInstance.isPresent()) {
+        continue;
+      }
+      final Optional<RunState> runState = lookupPodRunState(pod, workflowInstance.get());
+      if (runState.isPresent()) {
+        emitPodEvents(Watcher.Action.MODIFIED, pod, runState.get());
+      } else {
+        LOG.info("Deleting unwanted pod {}", pod.getMetadata().getName());
+        client.pods().delete(pod);
```

Comment:
Not too bad I think. As a bonus we now cache the debug flag and do not inject the entire `Storage` interface into the `DockerRunnerHandler`.